opnode/rollup/sync/start.go

```diff
@@ -9,102 +9,140 @@ import (
+		l1header, err := source.L1NodeByNumber(ctx, n.l1parent.Number)
 		if err != nil {
-			// If refL1 is the head block, then we might not have a next block to build on the head
-			if errors.Is(err, ethereum.NotFound) {
-				// return the same as the engine head was already built on, no error.
-				nextRefL1 = refL1
-				refL2 = eth.BlockID{Hash: parentL2, Number: refL2.Number}
-				if refL2.Number > 0 {
-					refL2.Number -= 1
-				}
-				err = nil
-				return
+			// Generic error, bail out.
+			if !errors.Is(err, ethereum.NotFound) {
+				return L2Node{}, fmt.Errorf("failed to fetch L1 block %v: %w", n.l1parent.Number, err)
+			}
+			// L1 block not found, keep walking chain
+		} else {
+			// L1 Block found, check if matches & should keep walking the chain
+			if l1header.self.Hash == n.l1parent.Hash {
+				return n, nil
 			}
-			return
 		}
-		// The L1 source might rug us with a reorg between API calls, catch that.
-		if ontoL1 != currentL1 {
-			err = fmt.Errorf("the L1 source reorged, the block for N+1 %s doesn't have the previously fetched block N %s as parent, but builds on %s instead", nextRefL1, currentL1, ontoL1)
+
+		// Don't walk past genesis. If we were at the L2 genesis, but could not find the L1 genesis
+		// pointed to from it, we are on the wrong L1 chain.
+		if n.self.Hash == genesis.L2.Hash || n.self.Number == genesis.L2.Number {
+			return L2Node{}, WrongChainErr
 		}
-		return
-	}
 
-	// Search back: linear walk back from engine head. Should only be as deep as the reorg.
-	for refL2.Number > 0 {
-		// remember the canonical L1 block that builds on top of the L1 source block of the L2 parent block.
-		nextRefL1 = currentL1
-		refL1, refL2, parentL2, err = reference.RefByL2Hash(ctx, parentL2, genesis)
+		// Pull L2 parent for next iteration
+		n, err = source.L2NodeByHash(ctx, n.l2parent.Hash, genesis)
 		if err != nil {
-			// TODO: re-attempt look-up, now that we already traversed previous history?
-			err = fmt.Errorf("failed to lookup block %s in L2: %w", refL2, err) // refL2 is previous parentL2
-			return
+			return L2Node{}, fmt.Errorf("failed to fetch L2 block by hash %v: %w", n.l2parent.Hash, err)
 		}
-		// Check if L1 source has the block that derived the L2 block we are planning to build on
-		currentL1, _, err = reference.RefByL1Num(ctx, refL1.Number)
+	}
+}
+
+func findL1Range(ctx context.Context, source SyncReferenceV2, begin eth.BlockID) ([]eth.BlockID, error) {
+	if _, err := source.L1NodeByNumber(ctx, begin.Number); err != nil {
+		return nil, fmt.Errorf("failed to fetch L1 block %v %v: %w", begin.Number, begin.Hash, err)
+	}
+	// TODO: Check hash here (even if slightly redudant)
+	l1head, err := source.L1HeadNode(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch head L1 block: %w", err)
+	}
+	maxBlocks := uint64(100)
+	if l1head.self.Number-begin.Number <= maxBlocks {
+		fmt.Println("Capping max blocks")
+		maxBlocks = l1head.self.Number - begin.Number
+		fmt.Println(maxBlocks)
+	}
```

Comment:
No, but we may not get a full sequence window of blocks based on chain conditions.

Reply:
This was me being confused by the purpose of this function for no good reason!

But then I have another question: what happens if we do cap the number of blocks? Wouldn't things fail because we haven't fetched all the blocks since `begin`?

Reply:
We start from `begin`, so this is actually stopping us from going past the head block in the case that `begin` is closer to `head` than `100`.